### DISCUSSION TITLE
Design question: Using add-on products in scanning processes

### DISCUSSION THREAD

**Discussion by raphaelm** [pretix/pretix#2539] [Ideas]
We have a use-case related to scanning that we need to solve. It has been mentioned in #2433 before and other places as well.

The main use case is:

> An event has a main entrance ticket, as well as optional add-ons. For example, the main ticket is "conference entry", and the addons are "workshop A", and "workshop B". pretixSCAN should be used for access control for the workshops.

Currently, this would require attendees to have the tickets for the workshops on them, so they need to carry three QR codes around. It would be much more convenient if they could just show **one** barcode, the one of their main ticket, everywhere. Maybe it's even printed on their badge and they don't even need to pull it out.

## Approach A

Let's assume that each workshop has an individual check-in list that **only** includes the workshop products.

We could build the following behavior into the check-in process, either optionally, or by default: If no ticket is found with the scanned barcode, look instead for any tickets that have the scanned barcode as a parent. If there is none, return "invalid ticket". If there is exactly one, use that one for check-in.

This would work quite well and has the advantage that the check-in record will be created directly for the add-on position.

The main issue is: What do we do if we find multiple matches, i.e. the scanned ticket has multiple add-on products that are on our check-in list? We could just fail, which would be okay of the exact use-case above.

Or we could use the first one that has not been used, which would open this up for new use-cases. For example, you could then use this to implement a multi-ticket bundle where the main ticket has 10 sub-tickets and every time a new one is used until all are used up. However, this could lead to bad synchronization conflicts in offline mode if we implement it the straight-forward way, so we could probably only support that in online mode at all.

## Approach B

Let's assume that each workshop has an individual check-in list that includes the **main ticket**, but not the workshop tickets.

We could build a new rule into our check-in rule engine that basically says "ticket needs to include workshop A". So the lookup and check-in would still happen on the main product, but conditionally on the fact that there is a "workshop A" ticket included in the add-on products.

This would do away with the complications of approach A when there are multiple matching add-on products, since we just wouldn't care. Once the first is found, we'd consider the condition fulfilled and move on.

This approach requires more users to use our rules engine, which is hard to use for new users. At the same time, once you're using it, this approach would also make it easier to phrase rules like "regular attendees need to have individual addon-tickets for every workshop, but exhibitors get in everywhere". This exact use case would still be possible with approach A, though, too, but with approach B you could escalate to complex logic like "exhibitors always get in, regular attendees only get in before 9am if they bought the early-entry-addon".

However, we could not potentially support the multi-pass usecase from above later on, and it might be more confusing to debug in the backend if all check-ins are assigned to the main order position.

## General limitations

Both approaches require access information on the parent-child relationships of tickets. Therefore, both can only be implemented in online mode and in offline mode with full data availability. There is no sensible way of implementing them with signed barcodes, unless we can afford the barcodes to become reeally big.

Both approaches also require careful configuration of check-in lists, but I think we'll need to accept that.

Opinions? Any use cases or approaches I haven't considered?

**Comment by pc-coholic** [comment]
While I do appreciate the simplicity of Approach A, I am very much in favor of making our rule engine more powerful...

Perhaps a good middle-way could be to extend the rule engine and add a few "presets" that can be selected from a drop-down menu. That might also make usage easier for users that just want to explore how certain scenarios would work with rules engine.

**Reply by raphaelm** [reply]
But… don't you think it will be much nicer if the check-in :heavy_check_mark: signs in the backend are on the actual add-on-product, instead of tens of :heavy_check_mark: on the main product that only differ in check-in list?

Approach A also – with a further improvements that allow check-in of *all* matching child projects – opens up future possibilities for group tickets where only the main tickets needs to be scanned, e.g. when a school class visits a museum and only the teacher needs to scan their ticket? Or a family ticket where only the main ticket needs to be scanned and not all individual ones?

**Reply by raphaelm** [reply]
I actually disagree that option A has "simplicity", I believe option A is the "correct, but hard" solution and option B is the "simple and quick hack".

**Comment by stapelberg** [comment]
> The main issue is: What do we do if we find multiple matches, i.e. the scanned ticket has multiple add-on products that are on our check-in list? We could just fail, which would be okay of the exact use-case above.

pretix 4.12 contains the `addon_match` field (cool!), which [seems to just return an error (REASON_AMBIGUOUS)](https://github.com/pretix/pretix/blob/69b482849a6bd1f89f47db7cea75be09f420cbe2/src/pretix/api/views/checkin.py#L556) when more than one add-on is matched.

I would have expected pretix to check-in all add-ons in this case, as the expectation is that a person comes only once to pick up all the add-ons they ordered. Would that be possible to implement?

**Reply by raphaelm** [reply]
> I would have expected pretix to check-in all add-ons in this case, as the expectation is that a person comes only once to pick up all the add-ons they ordered. Would that be possible to implement?

It's complicated:tm:

- Problem 1: Our check-in system is quite flexible and in the general case, there is no guarantee that all add-ons **can** be checked in. For example, one add-on might have a rule that it can only be checked in on Mondays, and another add-on only before 9am. Or, for some reason, one of the add-ons already *has* been checked in. → The app would need to be able to report a mixed status that indicated that the check-in was partially successful.

- Problem 2: There are other use cases where "just check in all" is not good enough. A common use case for add-ons is a family ticket, where the main ticket represents the family and the add-ons represent the individual persons within the family. If the main ticket is scanned, you *might* want to present the user with a **selection** of which add-ons to check in, in case not everyone is present yet.

Either of these problems is solvable. But we made the choice that – at least for now – we don't **want** to solve them, for two reasons:

- After we opted for just returning AMBIGUOUS for now, we could solve our primary use case for this with a 50 line patch for the server and a 50 line patch for the clients. If we would have attempted to solve the two problems above, it would have required a major user interface redesign of (a) our Android app (b) our Desktop app (c) our iOS app (d) our web check-in app. We would not have made the deadline of some events who required this feature. We can always still do this, if we ever truly feel we need it, the path is not blocked.

- I'm not fully sold it's a good idea. Experience from *all kinds of events* shows that *in the majority of cases* our check-in apps are operated by people who have maybe 5 minutes of training on what it does and how to operate it. Often less. I've literally been in situations where I needed to explain the app to a crowd of 30 people starting their shift in 5 minutes. "The app turns red, yellow or green" is simple, "there might be a table with partial results in some cases" or "you will need to select/deselect the correct addons" is absolutely not simple.

However in **your case** I don't see why you would need it at all: If (a) everyone who comes to your check-in desk picks up all of their add-ons at once and (b) the only way to pick up add-ons is at the check-in desk, I believe you do not need to care about the check-in status of add-ons at all and the check-in of the main ticket is sufficient. The only reason to track check-in of addons would be to have a simpler access to the number "how many of X are not yet picked up", but that could possibly be solved on reporting/output level instead of data/input level just as well.

**Reply by raphaelm** [reply]
One more note, as shown in this example, the hard part about developing features in pretix core is all the edge cases.

If you limit yourself to a specific use case and can make specific assumptions, it would be almost trivial to implement a plugin that listens on the signal that a new checkin was created and then just writes additional checkin objects for the addon-on tickets without further checks ;)

**Reply by stapelberg** [reply]
Thanks for your reply!

> Problem 1: Our check-in system is quite flexible and in the general case, there is no guarantee that all add-ons can be checked in. For example, one add-on might have a rule that it can only be checked in on Mondays, and another add-on only before 9am. Or, for some reason, one of the add-ons already has been checked in. → The app would need to be able to report a mixed status that indicated that the check-in was partially successful.

This could be solved on the technical level by checking in all add-ons in a single transaction, and thereby either succeeding or failing the whole check-in. Not sure how much work that’d be with the current code, though.

> Problem 2: There are other use cases where "just check in all" is not good enough. A common use case for add-ons is a family ticket, where the main ticket represents the family and the add-ons represent the individual persons within the family. If the main ticket is scanned, you might want to present the user with a selection of which add-ons to check in, in case not everyone is present yet.

Yeah, this problem I don’t think you can solve if you say that selection menus are a no-go in the scanning app (which is totally fair, IMO). It would probably be easiest to keep using separate tickets when multiple people’s orders are represented by add-ons.

> However in your case I don't see why you would need it at all: If (a) everyone who comes to your check-in desk picks up all of their add-ons at once and (b) the only way to pick up add-ons is at the check-in desk, I believe you do not need to care about the check-in status of add-ons at all and the check-in of the main ticket is sufficient.

Yeah, I think you’re right. We probably can indeed just look at the main check-in.

> If you limit yourself to a specific use case and can make specific assumptions, it would be almost trivial to implement a plugin that listens on the signal that a new checkin was created and then just writes additional checkin objects for the addon-on tickets without further checks ;)

That’s good to keep in mind! Thanks for the suggestion.

**Reply by raphaelm** [reply]
> This could be solved on the technical level by checking in all add-ons in a single transaction, and thereby either succeeding or failing the whole check-in. Not sure how much work that’d be with the current code, though.

Oh, sure, that would be an option! But then you have a new problem: Now your attendee can't be checked in *at all* just because one of the sub-tickets is in an unexpected state, with no way to move forward for the check-in staff.
